src/Uno.UI/UI/Xaml/UIElement.Pointers.cs

```diff
@@ -727,6 +728,17 @@ private async Task<DataPackageOperation> StartDragAsyncCore(PointerPoint pointer
 			PrepareShare(routedArgs.Data); // Gives opportunity to the control to fulfill the data
 			SafeRaiseEvent(DragStartingEvent, routedArgs); // The event won't bubble, cf. PrepareManagedDragAndDropEventBubbling
 
+			// We need to give a chance for  for layout updates, etc. This is particularly problematic with TreeView
+			// dragging where the DragStarting event on the TreeView will "internally" collapse some nodes,
+			// but actually removing them from the visual tree needs a layout cycle. Without waiting here,
+			// we can also get a DragEnter event on one of the to-be-collapsed containers in the same
+			// pointer event. This crashes the dragging logic.
+			// The way WinUI asynchronously handles StartDragAsync (relevant code is in AutomaticDragHelper::AutomaticDragHelper::HandlePointerMovedEventArgs)
+			// is not very clear and the way it interacts with layout timing is not similar at all to
+			// what we do in Uno, so the closest thing is to wait for Low/Idle here.
+			// Note that waiting for Idle causes some unrelated problems in ListView dragging, so we choose Low.
+			await NativeDispatcher.Main.EnqueueAsync(() => { }, NativeDispatcherPriority.Low);
```

Comment:
> the DragStarting event on the TreeView will "internally" collapse some nodes
FYI, we are handling the visual tree modification [using the `PointerEventDispatchResult`](https://github.com/unoplatform/uno/blob/9c24c59ce294f335a7c6884e45d8415638fc9f36/src/Uno.UI/UI/Xaml/UIElement.Pointers.cs#L243) I think this is similar to what you are facing here (but it's for managed pointers only).

> we can also get a DragEnter event on one of the to-be-collapsed containers
Awaiting a somehow random delay here does not make sure that it won't be the case anymore, especially if we have "recursed" updates (like altering again the visual tree in `OnLoaded` or `OnUnloaded` of one of your containers). At some point I would simply invoke `UIElement.UpdateLayout`. Is it something you already tried?

Reply:
Yes, `UpdateLayout` was already a fix I added in https://github.com/unoplatform/uno/pull/16762 (and forgot about it :/) but it looked very hacky to me. The delay here is not random. It's a delay with low enough priority to allow an invalidated visual tree to get layouted.

On WinUI, if you subscribe to `DragOver` on the dragging target and then do something like this on the dragging source:
```csharp
private void UIElement_OnDragStarting(UIElement sender, DragStartingEventArgs args)
    {
        Console.WriteLine("dragstarting");
        e.InvalidateMeasure(); // e is a UIElement in the visual tree with an overriden MeasureOverride that prints "measured"
        DispatcherQueue.TryEnqueue(DispatcherQueuePriority.High, () => Console.WriteLine("high"));
        DispatcherQueue.TryEnqueue(DispatcherQueuePriority.Normal, () => Console.WriteLine("normal"));
        DispatcherQueue.TryEnqueue(DispatcherQueuePriority.Low, () =>
        {
            Console.WriteLine("low");
            DispatcherQueue.TryEnqueue(DispatcherQueuePriority.Low, () => Console.WriteLine("lowlow"));
        });
    }
```

When you drag an element, you get this:
```
dragstarting
high
normal
low
measured
lowlow
dragover // this comes from the dragging target
dragover // this comes from the dragging target
dragover // this comes from the dragging target
dragover // this comes from the dragging target
...
```

So it's not only that it's updating the visual tree. It's also allowing any operations enqueued on the dispatcher to go through first.
> FYI, we are handling the visual tree modification [using the PointerEventDispatchResult](https://github.com/unoplatform/uno/blob/9c24c59ce294f335a7c6884e45d8415638fc9f36/src/Uno.UI/UI/Xaml/UIElement.Pointers.cs#L243) I think this is similar to what you are facing here (but it's for managed pointers only).

Totally forgot about that one, thanks! But hittesting is not the problem here. The problem is NOT that we incorrectly hit-test an element that's no longer in the visual tree. The problem is that the collapsed element is still in the tree even though the TreeView removed the node from its internal representation. This is fine and expected because the visual tree will be updated to match this internal representation on the next `MeasureOverride`. This is all behaving as intended. It only becomes problematic when you can a SINGULAR PointerMove event that both:
1. fires a DragStarting event on the dragging source (i.e. this PointerMove event is just enough to move out of the "drag box". This causes the TreeView to collapse and remove the dragged TreeViewItem and all its children from its internal representation. 
2. also fires a DragEnter+DragOver on one of the TreeViewItems that was just "collapsed internally". The callback of DragOver in TreeViewItem still assumes that it (and some other nodes) are still in the internal representation of the parent TreeView.

So the fix is to somehow leave some breathing room after 1. so that 2. no longer happens. The closest fix I could find is the one in this PR.

Reply:
> but it looked very hacky to me

This is not a hack, it's instead a good way to stabilize the tree after an event has been raised and user can have updated the tree. And BTW this is how the `ItemsRepeater` works in some circumstances.

> But hittesting is not the problem here

Yes, it was only an option to have a way to defer only if needed.


And anyway looking at your "events log" on WinUI, you change is probably the most appropriate indeed.